WordPress/gutenberg
# Entity-aware type signature for getEntityRecord and getEntityRecords

## What problem does this PR solve?

A subset of https://github.com/WordPress/gutenberg/pull/39025

This PR proposes a type signature for the `getEntityRecord` and `getEntityRecords` selectors that supports the following use-cases:

```ts
const commentDefault = getEntityRecord( {} as State, 'root', 'comment', 15 );
// commentDefault is Comment<'edit'>

const commentView = getEntityRecord( {} as State, 'root', 'comment', 15, {
	context: 'view',
} );
// commentView is Comment<'view'>

const commentInvalidPK = getEntityRecord(
	{} as State,
	'root',
	'comment',
	'15'
);
// commentInvalidPK shows a TypeScript error

const commentView2 = getEntityRecord( {} as State, 'root', 'comment', 15, {
	context: 'view',
	_fields: [ 'id' ],
} );
// commentView is Partial< Comment<'view'> >
```

And analogous use-cases for `getEntityRecords`.

## Testing Instructions

* Confirm the tests are green, except for what I mentioned above
* Confirm the types work as described above
* Confirm the types read well and make sense

cc @dmsnell @sarayourfriend @sirreal
